drivers/can/can_mcp25xxfd.c

```diff
@@ -0,0 +1,1663 @@
+		MCP25XXFD_REG_CON_B3, reg_value, MCP25XXFD_CON_B3_REQOP_MASK,
+		MCP25XXFD_REG_CON_B2, reg_value_validate, MCP25XXFD_CON_B2_OPMOD_MASK,
+		MODE_CHANGE_TIMEOUT_USEC, MODE_CHANGE_RETRIES, true);
+}
+
+static int mcp25xxfd_set_mode(const struct device *dev, can_mode_t mode)
+{
+	struct mcp25xxfd_data *dev_data = dev->data;
+
+	if (dev_data->started) {
+		return -EBUSY;
+	}
+
+	/* todo: Add CAN_MODE_ONE_SHOT support */
+	if ((mode & (CAN_MODE_3_SAMPLES | CAN_MODE_ONE_SHOT)) != 0) {
+		return -ENOTSUP;
+	}
+
+	if ((mode & CAN_MODE_NORMAL) != 0) {
+		dev_data->mcp25xxfd_mode = MCP25XXFD_OPMODE_NORMAL_CAN2;
+	}
+
+	if ((mode & CAN_MODE_FD) != 0) {
+#if defined(CONFIG_CAN_FD_MODE)
+		dev_data->mcp25xxfd_mode = MCP25XXFD_OPMODE_NORMAL_CANFD;
+#else
+		return -ENOTSUP;
+#endif
+	}
+
+	if ((mode & CAN_MODE_LISTENONLY) != 0) {
+		dev_data->mcp25xxfd_mode = MCP25XXFD_OPMODE_LISTEN_ONLY;
+	}
+
+	if ((mode & CAN_MODE_LOOPBACK) != 0) {
+		dev_data->mcp25xxfd_mode = MCP25XXFD_OPMODE_EXT_LOOPBACK;
+	}
+
+	dev_data->mode = mode;
+
+	return 0;
+}
+
+static int mcp25xxfd_set_timing(const struct device *dev, const struct can_timing *timing,
+				bool is_nominal)
+{
+	struct mcp25xxfd_data *dev_data = dev->data;
+	struct mcp25xxfd_nbtcfg *nominal_timing_reg;
+	struct mcp25xxfd_dbtcfg *data_timing_reg;
+	struct mcp25xxfd_tdc *tdc;
+	uint8_t *sjw_cached;
+	uint16_t addr_timing_reg, reg_size;
+	int ret;
+
+	if (!timing) {
+		return -EINVAL;
+	}
+
+	/* we have to be in a configuration mode at this point */
+	if (dev_data->started) {
+		return -EBUSY;
+	}
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	sjw_cached = is_nominal ? &dev_data->sjw : &dev_data->sjw_data;
+	if (timing->sjw != CAN_SJW_NO_CHANGE) {
+		*sjw_cached = timing->sjw;
+	}
+
+	addr_timing_reg = is_nominal ?  MCP25XXFD_REG_NBTCFG : MCP25XXFD_REG_DBTCFG;
+
+	if (is_nominal) {
+		nominal_timing_reg = mcp25xxfd_get_spi_buf_ptr(dev, sizeof(*nominal_timing_reg));
+		nominal_timing_reg->BRP = timing->prescaler - 1;
+		nominal_timing_reg->TSEG1 = timing->prop_seg + timing->phase_seg1 - 1;
+		nominal_timing_reg->TSEG2 = timing->phase_seg2 - 1;
+		nominal_timing_reg->SJW = *sjw_cached - 1;
+		reg_size = sizeof(*nominal_timing_reg);
+	} else {
+		data_timing_reg = mcp25xxfd_get_spi_buf_ptr(dev, sizeof(*data_timing_reg));
+		data_timing_reg->BRP = timing->prescaler - 1;
+		data_timing_reg->TSEG1 = timing->prop_seg + timing->phase_seg1 - 1;
+		data_timing_reg->TSEG2 = timing->phase_seg2 - 1;
+		data_timing_reg->SJW = *sjw_cached - 1;
+		reg_size = sizeof(*data_timing_reg);
+	}
+
+	ret = mcp25xxfd_write_reg(dev, addr_timing_reg, reg_size);
+	if (ret < 0) {
+		LOG_ERR("Failed to write device configuration [%d]", ret);
+		goto done;
+	}
+
+	tdc = mcp25xxfd_get_spi_buf_ptr(dev, sizeof(*tdc));
+	if (is_nominal) {
+		tdc->TDCMOD = MCP25XXFD_TDCMOD_DISABLED;
+	} else {
+		tdc->TDCMOD = MCP25XXFD_TDCMOD_AUTO;
+		tdc->TDCO = timing->prescaler * (timing->prop_seg + timing->phase_seg1);
+	}
```

Comment:
good idea, I'll update.

Reply:
It seems that the loopback mode requires transmitter delay compensation to be disabled. Otherwise, the device goes into a bus-off state during FD messages.

This was unintentionally done during the tests because they contain several calls to can_set_timing() that disabled the TDCO. 
The tests would have actually failed if there were any extra calls to can_set_timing_data(). This is not done in the API test yet (at least can_set_timing_data() that is supposed to succeed).

I've fixed this issue in the latest version of the driver. The TDCO is enabled/disabled when changing the internal mode.